### PR TITLE
python3Packages.apache-airflow init 1.10.5

### PR DESCRIPTION
**Pull request by ingenieroariel**
###### Motivation for this change
Apache Airflow is a widely used workflow tool. https://github.com/apache/airflow

Two PRs precede this one:
 -  #44028 Stalled because it had a really long list of dependencies and did not reuse existing python packages
 - #65026 Stalled because handling of py27 and py3k at the same time required a lot of decision making and touched on important Python ecosystem decisions not yet made: https://discourse.nixos.org/t/scaling-the-python-package-set-in-nixpkgs/3749

This PR builds on #65026 but avoids the py27 question completely by making it py3k only and patching airflow to be able to use the version of pendulum already in nixpkgs.

###### Things done

 - Squashed #65026 onto one commit and the update from 1.10.3 to 1.10.5 onto another commit.
 - Got rid of all import path rewriting and py27 conditionals for apache-airflow

- [ x ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ x ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ x ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/development/python-modules/apache-airflow/pendulum-fix.patch
Comment (flokli) [inline pkgs/development/python-modules/apache-airflow/pendulum-fix.patch]: can this be converted to a PR upstream (if it isn't already), and `fetchpatch`-ed from there?

File: pkgs/development/python-modules/apispec/default.nix
Comment (flokli) [inline pkgs/development/python-modules/apispec/default.nix]: you can use `pytestCheckHook` instead.

File: pkgs/development/python-modules/flask-admin/default.nix
Comment (flokli) [inline pkgs/development/python-modules/flask-admin/default.nix]: instead of removing these files, can you pass them via `-I` to `nosetests`?

File: pkgs/development/python-modules/flask-appbuilder/default.nix
Comment (flokli) [inline pkgs/development/python-modules/flask-appbuilder/default.nix]: instead of removing these files, can you pass them via `-I` to `nosetests`?

File: pkgs/development/python-modules/flask-appbuilder/default.nix
Comment (flokli) [inline pkgs/development/python-modules/flask-appbuilder/default.nix]: so there's more tests that would need to be disabled? In that case, it might make sense to remove `checkPhase` entirely…

File: pkgs/development/python-modules/flask-babelex/default.nix
Comment (flokli) [inline pkgs/development/python-modules/flask-babelex/default.nix]: why were these skipped? maybe add a comment.

File: pkgs/development/python-modules/flask-bcrypt/default.nix
Comment (flokli) [inline pkgs/development/python-modules/flask-bcrypt/default.nix]: we might be able to just say `python test_bcrypt.py` here.

File: pkgs/development/python-modules/json-merge-patch/default.nix
Comment (flokli) [inline pkgs/development/python-modules/json-merge-patch/default.nix]: is this a worth a patch upstream?

File: pkgs/development/python-modules/marshmallow-enum/default.nix
Comment (flokli) [inline pkgs/development/python-modules/marshmallow-enum/default.nix]: Use `pytestCheckHook` instead, maybe setting `disabledTests` instead of patching the `mark_xfail` in. Why does it fail?

File: pkgs/development/python-modules/sqlalchemy-citext/default.nix
Comment (flokli) [inline pkgs/development/python-modules/sqlalchemy-citext/default.nix]:
```suggestion
   python tests/test_citext.py
```

File: pkgs/development/python-modules/wtf-peewee/default.nix
Comment (flokli) [inline pkgs/development/python-modules/wtf-peewee/default.nix]:
```suggestion
    python runtests.py
```

**Review by flokli — CHANGES_REQUESTED** [review]
Thanks a lot, must have been a lot of work! Some suggestions.

**Comment by ingenieroariel** [comment]
> Thanks a lot, must have been a lot of work! Some suggestions.

All the credit for the approach and hard work goes to @costrouc

I'll unify the approach on skipping tests / fetching patches to upstream from PRs created to their repo. Thanks a lot for the quick feedback @flokli!

**Comment by ingenieroariel** [comment]
> Thanks a lot, must have been a lot of work! Some suggestions.

All the credit for the approach and hard work goes to @costrouc

I'll unify the approach on skipping tests / fetching patches to upstream from PRs created to their repo. Thanks a lot for the quick feedback @flokli!

**Comment by ingenieroariel** [comment]
> Thanks a lot, must have been a lot of work! Some suggestions.

All the credit for the approach and hard work goes to @costrouc

I'll unify the approach on skipping tests / fetching patches to upstream from PRs created to their repo. Thanks a lot for the quick feedback @flokli!

File: pkgs/development/python-modules/wtf-peewee/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/wtf-peewee/default.nix]: ${python.interpreter} is correct, it allows for other interpreters such as pypy

File: pkgs/development/python-modules/sqlalchemy-citext/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/sqlalchemy-citext/default.nix]: again, this is considered best pratice to allow for pypy and others

File: pkgs/development/python-modules/flask-bcrypt/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/flask-bcrypt/default.nix]: for other interpreters, this is best practice

File: pkgs/development/python-modules/apache-airflow/pendulum-fix.patch
Comment (ingenieroariel) [reply pkgs/development/python-modules/apache-airflow/pendulum-fix.patch]: There was one that was stalled: https://github.com/apache/airflow/pull/4507

I will create a new one.

Comment (ingenieroariel) [reply pkgs/development/python-modules/apache-airflow/pendulum-fix.patch]: Created: https://github.com/apache/airflow/pull/6561#issuecomment-553072328

File: pkgs/development/python-modules/json-merge-patch/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/json-merge-patch/default.nix]: Left a patch there and pulled it back via fetchpatch: https://github.com/OpenDataServices/json-merge-patch/pull/1

File: pkgs/development/python-modules/flask-appbuilder/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/flask-appbuilder/default.nix]: Did exactly that.

**Comment by ingenieroariel** [comment]
@flokli I think I covered all the comments/suggestions with the new commit. If I missed anything please let me know. Thanks in advance!

File: pkgs/development/python-modules/sqlalchemy-citext/default.nix
Comment (flokli) [reply pkgs/development/python-modules/sqlalchemy-citext/default.nix]: Right.

File: pkgs/development/python-modules/wtf-peewee/default.nix
Comment (flokli) [reply pkgs/development/python-modules/wtf-peewee/default.nix]: Right.

File: pkgs/development/python-modules/flask-babelex/default.nix
Comment (flokli) [inline pkgs/development/python-modules/flask-babelex/default.nix]: sounds a bit like the time-related translations aren't applied properly, or have been updated since the tests were written. I couldn't find them in the `translations` subfolder. Did you check if airflow l10n in general works? If it supports it, at all :-D

**Review by flokli — APPROVED** [review]
This already looks quite good! Thanks a lot. @FRidh, could you take a look as well?

**Comment by flokli** [comment]
Not something that should stall this PR, but @ingenieroariel do you have a NixOS module in the pipeline aswell, or some quick notes on how to get it running?

It might be beneficial to have some NixOS tests that spins up airflow, and does some curl request to see it was able to do the initialization.

**Comment by ingenieroariel** [comment]
Thanks @flokli I have not yet thought about the module but it makes a lot of sense to get started on that after this gets merged.

Copying people from previous PRs who expressed interest in this being merged:
cc @offlinehacker @costrouc @cawilliamson @volth @risicle @costrouc @Ericson2314

File: pkgs/development/python-modules/apispec/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/apispec/default.nix]: I don't think you meant to add costrouc as a maintainer

File: pkgs/development/python-modules/apache-airflow/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/apache-airflow/default.nix]: do you mind adding a comment on what the patches due? and if relevant, if they should be removed on the next bump.

File: pkgs/development/python-modules/flask-mongoengine/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/flask-mongoengine/default.nix]: I would remove costrouc as a maintainer, unless he has your permission to do so. But I would l ike to see you add yourself :)

File: pkgs/development/python-modules/flask-openid/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/flask-openid/default.nix]: a lot of packages dont export tests to pypi, may want to checkout from github

File: pkgs/development/python-modules/json-merge-patch/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/json-merge-patch/default.nix]: could you explain why the patch is necessary in a comment? :)

File: pkgs/development/python-modules/json-merge-patch/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/json-merge-patch/default.nix]: should probably remove costrouc as a maintainer

File: pkgs/development/python-modules/mongoengine/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/mongoengine/default.nix]: should probably remove costrouc as a maintainer

File: pkgs/development/python-modules/prance/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/prance/default.nix]: should probably remove costrouc as a maintainer

File: pkgs/development/python-modules/prance/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/prance/default.nix]: shouldn't be necessary since you skip the checkPhase anyway

File: pkgs/development/python-modules/prance/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/prance/default.nix]: there probably some unit tests that you could run

File: pkgs/development/python-modules/python-openid/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/python-openid/default.nix]: you can do `nosetests -I=file_that_needs_networking` to ignore specific files, would still like to see at least a few tests to ensure teh package is roughly usable.

File: pkgs/development/python-modules/prison/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/prison/default.nix]: should probably remove costrouc as a maintainer

File: pkgs/development/python-modules/python-openid/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/python-openid/default.nix]: should probably remove costrouc as a maintainer

File: pkgs/development/python-modules/snakebite/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/snakebite/default.nix]: should probably remove costrouc as a maintainer

File: pkgs/development/python-modules/snakebite/default.nix
Comment (jonringer) [inline pkgs/development/python-modules/snakebite/default.nix]: I think argparse is still needed for python2, might be wrong.

**Comment by ingenieroariel** [comment]
@jonringer He created the package as part of https://github.com/NixOS/nixpkgs/pull/65026 and that addition is on a commit made by him that this PR respects.

I could add myself too if that is best practice (I did as much in the main airflow package) but it felt rude to just retroactively claim ownership.

Still want me to add myself? No problem doing so as I want to keep airflow working on unstable.

**Comment by jonringer** [comment]
I guess I must have overlooked that this was based upon a previous PR, keeping him there is fine, but feel free to add yourself

**Comment by jonringer** [comment]
I resolved the issues dealing with maintainership

**Comment by jonringer** [comment]
Generally when adding python-modules, it's roughly one commit = one package addition. @FRidh how strictly do you want to adhere to that role, I think there's 18 or 19 packages being added

**Comment by ingenieroariel** [comment]
> Generally when adding python-modules, it's roughly one commit = one package addition. @FRidh how strictly do you want to adhere to that role, I think there's 18 or 19 packages being added

I squashed the commits to make porting to latest master easier, each package folder is fine but the `top-level/python-packages.nix` file is problematic when a new package in master falls in the middle of two packages from this branch. If it is strictly needed I can re-create from scratch on top of master but commits may loose correct ownership.

**Comment by offlinehacker** [comment]
I am not sure with a way this package is packaged, since you override all version requirements, how can you be sure things will work. I did a pull request a while back that used pypi2nix, it generated a large dependency file, but at least I was sure dependencies are right.

File: pkgs/development/python-modules/flask-openid/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/flask-openid/default.nix]: I checked on github and it seems there really are no tests: https://github.com/mitsuhiko/flask-openid/commit/83534279b9f10e615612b837738d2817b01bcf1a

File: pkgs/development/python-modules/apache-airflow/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/apache-airflow/default.nix]: Did this:

```
   patches = [
+       # Not yet accepted: https://github.com/apache/airflow/pull/6562
      (fetchpatch {
+       name = "avoid-warning-from-abc.collections";
        url = https://patch-diff.githubusercontent.com/raw/apache/airflow/pull/6562.patch;
        sha256 = "0swpay1qlb7f9kgc56631s1qd9k82w4nw2ggvkm7jvxwf056k61z";
      })
+       # Not yet accepted: https://github.com/apache/airflow/pull/6561
      (fetchpatch {
+       name = "pendulum2-compatibility"; 
        url = https://patch-diff.githubusercontent.com/raw/apache/airflow/pull/6561.patch;
        sha256 = "17hw8qyd4zxvib9zwpbn32p99vmrdz294r31gnsbkkcl2y6h9knk";

```

File: pkgs/development/python-modules/json-merge-patch/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/json-merge-patch/default.nix]: Did this:
```
   patches = [
+    # This prevented tests from running (was using a relative import)
+    # https://github.com/OpenDataServices/json-merge-patch/pull/1
    (fetchpatch {
+     name = "fully-qualified-json-merge-patch-import-on-tests";
      url = https://patch-diff.githubusercontent.com/raw/OpenDataServices/json-merge-patch/pull/1.patch;
```

File: pkgs/development/python-modules/python-openid/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/python-openid/default.nix]: > I am not sure with a way this package is packaged, since you override all version requirements, how can you be sure things will work. I did a pull request a while back that used pypi2nix, it generated a large dependency file, but at least I was sure dependencies are right.

This is a good point, I commented below the actual version we ended up choosing (based on what is available in NixOS):

```
--replace "flask>=1.1.0, <2.0" "flask" \  # got 1.1.1
--replace "flask-caching>=1.3.3, <1.4.0" "flask-caching" \ # got 1.7.2
--replace "flask-appbuilder>=1.12.5, <2.0.0" "flask-appbuilder" \ # got 2.1.6
--replace "pendulum==1.4.4" "pendulum" \ # got 2.0.5
--replace "cached_property~=1.5" "cached_property" \ # got 1.5.1
--replace "dill>=0.2.2, <0.3" "dill" \ # go 0.3.1.1
--replace "configparser>=3.5.0, <3.6.0" "configparser" \ # got 4.0.2
--replace "jinja2>=2.7.3, <=2.10.0" "jinja2" \ # got 2.10.3
--replace "funcsigs==1.0.0" "funcsigs" \ # got 1.0.2
--replace "flask-swagger==0.2.13" "flask-swagger" \ # got 0.2.14
--replace "python-daemon>=2.1.1, <2.2" "python-daemon" \ # got 2.2.3
--replace "alembic>=0.9, <1.0" "alembic" \ # got 1.2.1
--replace "markdown>=2.5.2, <3.0" "markdown" \ # 3.1.1
--replace "future>=0.16.0, <0.17" "future" \ # got 0.18.1
--replace "tenacity==4.12.0" "tenacity" \ # # got 5.1.1
--replace "text-unidecode==1.2" "text-unidecode" \ # got 1.3
--replace "tzlocal>=1.4,<2.0.0" "tzlocal" \ # got 2.0.0
--replace "sqlalchemy~=1.3" "sqlalchemy" \ # got 1.3.10
--replace "werkzeug>=0.14.1, <0.15.0" "werkzeug" # got 0.16.0
```

I do not expect markdown, sqalchemy, werkzeug or jinja2 to make wild changes on minor updates, but for those apps that made incompatible changes (like pendulum), patches to airflow have been provided. setup.py may look strict but from experience it is always a judgement call on the last person who created a release: in our case, the nixos package maintainer. After that, we only have the test suite to protect us and hopefully some module tests in the near future for integration testing.

**Comment by ingenieroariel** [comment]
> I am not sure with a way this package is packaged, since you override all version requirements, how can you be sure things will work. I did a pull request a while back that used pypi2nix, it generated a large dependency file, but at least I was sure dependencies are right.

This is a good point, I commented below the actual version we ended up choosing (based on what is available in NixOS):

```
--replace "flask>=1.1.0, <2.0" "flask" \  # got 1.1.1
--replace "flask-caching>=1.3.3, <1.4.0" "flask-caching" \ # got 1.7.2
--replace "flask-appbuilder>=1.12.5, <2.0.0" "flask-appbuilder" \ # got 2.1.6
--replace "pendulum==1.4.4" "pendulum" \ # got 2.0.5
--replace "cached_property~=1.5" "cached_property" \ # got 1.5.1
--replace "dill>=0.2.2, <0.3" "dill" \ # go 0.3.1.1
--replace "configparser>=3.5.0, <3.6.0" "configparser" \ # got 4.0.2
--replace "jinja2>=2.7.3, <=2.10.0" "jinja2" \ # got 2.10.3
--replace "funcsigs==1.0.0" "funcsigs" \ # got 1.0.2
--replace "flask-swagger==0.2.13" "flask-swagger" \ # got 0.2.14
--replace "python-daemon>=2.1.1, <2.2" "python-daemon" \ # got 2.2.3
--replace "alembic>=0.9, <1.0" "alembic" \ # got 1.2.1
--replace "markdown>=2.5.2, <3.0" "markdown" \ # 3.1.1
--replace "future>=0.16.0, <0.17" "future" \ # got 0.18.1
--replace "tenacity==4.12.0" "tenacity" \ # # got 5.1.1
--replace "text-unidecode==1.2" "text-unidecode" \ # got 1.3
--replace "tzlocal>=1.4,<2.0.0" "tzlocal" \ # got 2.0.0
--replace "sqlalchemy~=1.3" "sqlalchemy" \ # got 1.3.10
--replace "werkzeug>=0.14.1, <0.15.0" "werkzeug" # got 0.16.0
```

I do not expect markdown, sqalchemy, werkzeug or jinja2 to make wild changes on minor updates, but for those apps that made incompatible changes (like pendulum), patches to airflow have been provided. setup.py may look strict but from experience it is always a judgement call on the last person who created a release: in our case, the nixos package maintainer. After that, we only have the test suite to protect us and hopefully some module tests in the near future for integration testing.

**Comment by jonringer** [comment]
If the versions of the dependencies do need diverge from whats on nixpkgs, then you can move this out of python-modules and create an application. At least there you can wrap the program with it's own pythonpath.

Having more than 1 version in python-modules isn't acceptable due to how environments are created with python packages (both will be present, only 1 will get used)

**Comment by ingenieroariel** [comment]
> If the versions of the dependencies do need diverge from whats on nixpkgs, then you can move this out of python-modules and create an application. At least there you can wrap the program with it's own pythonpath.
> 
> Having more than 1 version in python-modules isn't acceptable due to how environments are created with python packages (both will be present, only 1 will get used)

This PR does not add more than one version of any packages, it attempts to reuse what is there. Both a top level application and a python-module are included here. For my project I do need a python-module on NixOS.

File: pkgs/development/python-modules/python-openid/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/python-openid/default.nix]: This module was completely removed as it was just adding python2Packages.python-openid which is not needed for this python3Packages.apache-airflow.

File: pkgs/development/python-modules/snakebite/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/snakebite/default.nix]: argparse is part of Python 2.7:

https://docs.python.org/2/library/argparse.html

File: pkgs/development/python-modules/flask-babelex/default.nix
Comment (ingenieroariel) [reply pkgs/development/python-modules/flask-babelex/default.nix]: I have not checked airflow I10n but I opened a PR on their repo to follow up: https://github.com/mrjoes/flask-babelex/issues/17

**Comment by ingenieroariel** [comment]
Thanks all for the comments and reviews. Should I do a one package per commit and force push the branch or is this okay as is?

File: pkgs/development/python-modules/snakebite/default.nix
Comment (jonringer) [reply pkgs/development/python-modules/snakebite/default.nix]: i must have been thinking of pathlib or something else. My bad :)

**Comment by jonringer** [comment]
not really sure, I was hoping @FRidh would make a more opinionated stance

**Comment by ingenieroariel** [comment]
@jonringer I went ahead and rebased one commit per package onto latest master, preserving original autorship.

**Comment by jonringer** [comment]
do you mind editing your commit to be:

```
apache-airflow: 1.10.3 -> 1.10.5, enable python
```

**Comment by ingenieroariel** [comment]
> ```
> apache-airflow: 1.10.3 -> 1.10.5, enable python
> ```
Done.

**Comment by jonringer** [comment]
The  last commit, https://github.com/NixOS/nixpkgs/pull/73074/commits/7af3a955c968647e6ed92c55d93bb4d8cb26add1 touched around 10 different files, it should have been broken into ~9 commits on what packages it's bump, removing, or altering

**Comment by flokli** [comment]
@jonringer Urgh, you're right. Should we revert and do a new PR?

**Comment by jonringer** [comment]
not entirely sure, but if there's a need to cherrypick or remove a package change, then we will have to revert that commit

**Comment by jonringer** [comment]
reverting, then applying the 9 commits would probably be the most proper course of action

**Comment by flokli** [comment]
Sorry, lost track about this. I think reverting it causes even more noise. We probably won't backport things from here to 19.09, so it'll hopefully be fine…

**Comment by flokli** [comment]
Sorry, lost track about this. I think reverting it causes even more noise. We probably won't backport things from here to 19.09, so it'll hopefully be fine…

**Pull request merged**
